### ISSUE TITLE
NoClassDefFoundError: MethodAccessor

### ISSUE THREAD

**Issue by palexdev** [Moderocky/Mirror#4]
Hello, I'm trying to improve the performance of my app by trying alternatives to reflection. ReflectASM seems to be abandoned, but your project on the other hand is quite interesting.

I'm trying to instantiate an object like this:
```java
Insets insets0 = (Insets) Mirror.of(Insets.class)
    .constructor(double.class)
    .newInstance(10.0);
assert insets0 != null;

Insets insets1 = (Insets) Mirror.of(Insets.class)
    .constructor(double.class, double.class, double.class, double.class)
    .newInstance(1.0, 2.0, 3.0, 4.0);
assert insets1 != null;
```

For some reason, it ends up throwing this exception:
```
Exception in thread "main" java.lang.RuntimeException: Unable to load class.
	at mx.kenzie.mirror.InternalAccessProvider.loadClass(InternalAccessProvider.java:88)
	at mx.kenzie.mirror.LookingGlass.loadClass(LookingGlass.java:532)
	at mx.kenzie.mirror.LookingGlass.createAccessor(LookingGlass.java:52)
	at mx.kenzie.mirror.Mirror.constructor(Mirror.java:103)
	at reproducer.Main.main(Main.java:10)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at mx.kenzie.mirror.InternalAccessProvider.loadClass(InternalAccessProvider.java:85)
	... 4 more
Caused by: java.lang.NoClassDefFoundError: mx/kenzie/mirror/MethodAccessor
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.System$2.defineClass(System.java:2390)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 6 more
Caused by: java.lang.ClassNotFoundException: mx.kenzie.mirror.MethodAccessor
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	... 9 more
```

Here's the MRE project: [mirror-reproducer](https://www.dropbox.com/scl/fi/uifvzp77fum7o42t6e2vv/mirror-reproducer.zip?rlkey=j4uhanreme6e5rl205b9wxicg&st=db16r8k7&dl=0)

**Comment by Moderocky** [comment]
I'll take a look at it tomorrow. :)

Things have changed significantly since this was made (and it was quite precarious at the time!) so I'm not surprised it's not working anymore.

That said, if you'd like, I can take a look at your project and see if I can find you a way to save time in another way?

**Comment by palexdev** [comment]
> I'll take a look at it tomorrow. :)
> 
> Things have changed significantly since this was made (and it was quite precarious at the time!) so I'm not surprised it's not working anymore.
> 
> That said, if you'd like, I can take a look at your project and see if I can find you a way to save time in another way?

Thank you very much!

My project is at [ArchitectFX](https://github.com/palexdev/ArchitectFX).
I'm relying on [jOOR](https://github.com//jOOQ/jOOR) to hide the complexity of reflection, but in the end it's plain old reflection which is a bit slow for my use case. The app relies heavily on reflection to instantiate objects, set and get their properties, because such objects are not know at compile time (basically loads jars from Maven and a 'config' file tells how to create and instantiate them)
By profiling the app, it seems that the most heavy weight task is instantiation, done [here](https://github.com/palexdev/ArchitectFX/blob/2bb3fa03906b7df05db0268720fa264caf3b56ad/src/main/java/io/github/palexdev/architectfx/utils/reflection/Reflector.java#L43)

**Comment by palexdev** [comment]
Hey @Moderocky, great news, I think I know what is going wrong!

So, [jOOR](https://github.com/jOOQ/jOOR) is smart, in the sense that if a constructor is not found for the given arguments, it still tries to find one suitable among the declared ones. This functionality is crucial. Let me explain why by using my case: the deserialization uses wrapper types rather than primitives, which makes sense. So, my args for the `Insets` class are `[Double, Double, Double, Double]`, but the constructor to use accepts `[double, double, double, double]`.

What I've done to adapt to such situations using your library is this (mostly copied from jOOR):
```java
try {
  Class <?>[] types = Arrays.stream(args)
      .map(Object::getClass)
      .toArray(Class[]::new);

  try {
    return (T) Mirror.of(klass)
            .setLoader(dm.loader())
            .constructor(types)
            .newInstance(args);
  } catch(Exception ex) {
    for (Constructor <?>constructor: klass.getDeclaredConstructors()) {
      if (match(constructor.getParameterTypes(), types)) {
        return (T) Mirror.of(klass)
                .setLoader(dm.loader())
                .constructor(constructor)
                .newInstance(args);
      }
    }

    throw new ReflectException(ex);
  }

} catch(Throwable ex) {
  Logger.error("Failed to create class {}:\n{}", klass.getName(), ex);
  return null;
}
```

The issue appears when calling [this](https://github.com/Moderocky/Mirror/blob/d50b5b45154ad14bb09c1c6cbd2533376c8bf5b6/src/main/java/mx/kenzie/mirror/Mirror.java#L114) method of Mirror as it calls `target.getClass()` which in my case results in having `java.lang.Class` instead of `javafx.geometry.Insets`
I don't know if that's an oversight or it's intentional, or even if I'm wrong haha, let me know 😉

Edit: I can't even reproduce the NoClassDefFoundError expection anymore, for some reason 🤦‍♂️

Edit 2: if I add this to the reproducer I sent you, I can't reproduce the issue (with the new target being Class rather than Insets)
```java
Insets insets2 = (Insets) Mirror.of(Insets.class)
    .constructor(Insets.class.getDeclaredConstructors()[1])
    .newInstance(1.0, 2.0, 3.0, 4.0);
assertNotNull(insets2);
assertEquals(new Insets(1.0, 2.0, 3.0, 4.0), insets2);
```

Edit 3: I can confirm that by changing the Mirror's method to:
```java
public <Type> ConstructorAccessor<Type> constructor(Constructor<Type> constructor) {
    if (target instanceof Class<?> type)
        return glass.createAccessor(type, constructor);
    return glass.createAccessor(target.getClass(), constructor);
}
```
`Insets` objects are now successfully created.

**Comment by Moderocky** [comment]
I'm glad you found a solution!

Interestingly enough, your reproducer worked absolutely fine for me.
<img width="760" alt="image" src="https://github.com/user-attachments/assets/86ad18cf-2ff2-44cb-86db-02c2d1f5cd7f">
<img width="422" alt="image" src="https://github.com/user-attachments/assets/9f77382d-d1b3-40a9-b037-0d253ee244bd">

It's possible I may have fixed something in my local copy in the years since and just not published it? I thought unboxing was handled internally but I could be wrong.
I notice you used `5.0.3` as the dependency,
<img width="356" alt="image" src="https://github.com/user-attachments/assets/3d8fcdbc-2b44-40b1-9ad1-17debe4cb716">
but I think there's a version `5.0.5`, not sure if that had any changes to this but it might be a factor.

If you think you have a way to make it more stable and would like to open a pull request with your changes, you're more than welcome, or I can tweak things on my end and see what I can do.

Other notes about reflection
=====

Reflection of this kind is, in essence, not actually that slow.

Basically it writes the _actual_ code you need while the program is running, and loads that as its own class.
The way Mirror wrote the code was slightly more efficient than the JDK's reflection classes (and I got to bypass a bunch of the security checks it does) which makes things a little faster, but in essence neither way is actually that slow.

The slowest part is looking up the method/constructor definitions and writing that "bridging" code in the first place. My biggest recommendation for your project is to make your `Constructor`s or `ConstructorAccessor`s in advance, and **keep hold of them** wherever possible, so you can reuse them rather than having to make new ones. :)

**Comment by palexdev** [comment]
So, in the end, here's my experience:
1) The library is very unstable compared to plain reflection. Many times it just throws exceptions and fails to build objects.
2) I thought it would be a bit faster than reflection, but it really is not. I don't know if I should have implemented a cache mechanism to go with it.
3) Types unboxing is not automatic, which means that I have to "manually" convert them to get the correct constructor. Not really a big deal, but jOOR handles such cases and many more automatically, hiding complexity and resulting in a more readable code

I'm still going to use jOOR as it is more stable, should automatically cache accessors, and overall the API is more fluent/convenient.
I sped up my app by parallelizing objects instantiations with virtual threads and I don't think I can optimize it further

Thanks for the support anyway, have a nice day!
